SQL/advanced/triggers/02_before_after.sql

```diff
@@ -0,0 +1,140 @@
+/*******************************************
+Difference between BEFORE and AFTER triggers
+*******************************************/
+
+DROP FUNCTION IF EXISTS
+  tr_delete_client, tr_search_name,
+  tr_clients_balances, tr_clients_standardize_name
+CASCADE;
+
+/********************
+BEFORE triggers
+********************/
+
+-- efficiency win (BEFORE just saves time not to do delete)
+CREATE FUNCTION tr_delete_client()
+  RETURNS trigger
+  LANGUAGE plpgsql
+AS $$
+/**
+Check client's status before deletion
+*/
+BEGIN
+  IF OLD.status = 'VIP' THEN
+    RAISE EXCEPTION 'The precious client % cannot leave our company!', OLD.id;
+  END IF;
+
+  RETURN OLD;
+END;
+$$;
+
+CREATE TRIGGER client_tr_delete
+  BEFORE DELETE ON clients
+  FOR EACH ROW
+  EXECUTE PROCEDURE tr_delete_client();
+
+
+-- the need for modification before UPDATE / INSERT
+CREATE FUNCTION tr_clients_standardize_name()
+  RETURNS trigger
+  LANGUAGE plpgsql
+AS $$
+DECLARE
+  _name_array text[];
+
+BEGIN
+  _name_array := string_to_array(NEW.name, ' ');
+
+  -- check for format
+  IF array_length(_name_array, 1) <> 2 THEN
+    RAISE EXCEPTION 'Wrong name: %', NEW.name;
+  END IF;
+
+  -- format the name correctly
+  NEW.name := initcap(NEW.name);
+
+  RETURN NEW;
+END;
+$$;
+
+CREATE TRIGGER clients_tr_standardize_name
+  BEFORE INSERT OR UPDATE OF name ON clients
+  FOR EACH ROW
+  EXECUTE PROCEDURE tr_clients_standardize_name();
+
+
+/********************
+AFTER triggers
+********************/
+
+-- more logical to do extra changes AFTER action
+CREATE FUNCTION tr_search_name()
+  RETURNS trigger
+  LANGUAGE plpgsql
+AS $$
+/**
+Update search table storing name ts_vectors for clients
+*/
+DECLARE
+  _ts tsvector;
+
+BEGIN
+  IF TG_OP = 'DELETE' THEN
+    DELETE FROM
+      search_clients
+    WHERE
+      object_id = OLD.id;
+
+    RETURN OLD;
+  END IF;
+
+  -- Weights and etc. are a bit different concept, let's be as simple as that :)
+  _ts := to_tsvector(NEW.name);
+
+  IF TG_OP = 'UPDATE' THEN
+    UPDATE
+      search_clients
+    SET
+      ts = _ts
+    WHERE
+      object_id = NEW.id;
```

Comment:
that's actually an interesting case.... try to think what can go wrong here with you approach.... what kind of query should I run, so that your trigger will do not what actually should have been expected...

Reply:
Do you mean updating an `id`? OK, this is very rare, but still actual...

What about, then, such a query:
```
UPDATE
  search_clients
SET
  ts = _ts,
  object_id = NEW.id
WHERE
  object_id = OLD.id;
```
?

Reply:
Sent a commit without a final "yes" as will not have access to a PC till tomorrow...

Reply:
Yep, correct.